python_modules/dagster/dagster/_core/instance/persist_run_for_production.py

```diff
@@ -0,0 +1,71 @@
+from typing import Mapping, Optional
+
+from dagster._core.definitions.utils import validate_tags
+from dagster._core.host_representation.repository_location import RepositoryLocation
+from dagster._core.host_representation.selector import PipelineSelector
+from dagster._core.instance import DagsterInstance
+from dagster._core.storage.pipeline_run import DagsterRun, DagsterRunStatus
+from dagster._core.utils import make_new_run_id
+
+
+def persist_run_for_production(
+    *,
+    instance: DagsterInstance,
+    repository_location: RepositoryLocation,
+    pipeline_selector: PipelineSelector,
+    run_config: Mapping[str, object],
+    context_specific_tags: Mapping[str, str],
+    explicit_mode: Optional[str],
+) -> DagsterRun:
+    """
+    Creates a run suitable for production using host process (i.e. External*) APIs.
+    This orchestrates necessary interactions with the user process (such as
+    fetching the external pipeline appropriate to the passed subset and
+    compiling the execution plan). This persists a run in instance with the
+    NOT_STARTED state.
+
+    Parameters:
+        instance (DagsterInstance): Instace to execute against
+        repository_location (RepositoryLocation): RepositoryLocation corresponding to user code
+        pipeline_selector (PipelineSelector): Selector that encapsulates subset of pipeline that will be executed
```

Comment:
It's kind of wrinkly that the repository location needs to be specified both in the `pipeline_selector` and the `repository_location`.

Reply:
Yeah not totally ideal but I think wrangling PipelineSelector is a separate workstream

Reply:
@sryza  just to follow up here I don't want to further contribute to the pattern of threading both `solid_selection` and `asset_selection` everywhere. I think we should have a single object that represents a subselection of the job and thread that through the entire codebase. I find this sort of repetitive passing of dual parameters that have implicit invariants between them to be pretty confusing.

Reply:
A single object that represents a subselection of a job sounds even better. Again not blocking feedback, but I think this could be a good place to introduce that - doesn't necessarily require wrangling the entire codebase.  Callers who want to use the new API and are currently working with a PipelineSelector could just do:

```python
persist_run(
    repository_name=my_pipeline_selector.repository_name,
    pipeline_name=my_pipeline_selector.pipeline_name,
    selection=my_pipeline_selector.selection,
    ...
)
```